core/src/main/scala/pureconfig/ConfigConvert.scala

```diff
@@ -56,76 +56,139 @@ trait ConfigConvert[T] {
 trait AllowMissingKey { self: ConfigConvert[_] => }
 
 object ConfigConvert extends LowPriorityConfigConvertImplicits {
+
   def apply[T](implicit conv: ConfigConvert[T]): ConfigConvert[T] = conv
 
-  private def fromFConvert[T](fromF: String => Try[T]): ConfigValue => Try[T] =
+  type ConfigReaderResult[A] = Either[ConfigReaderFailures, A]
+
+  def combineResults[A, B, C](first: ConfigReaderResult[A], second: ConfigReaderResult[B])(f: (A, B) => C): ConfigReaderResult[C] =
+    (first, second) match {
+      case (Right(a), Right(b)) => Right(f(a, b))
+      case (Left(aFailures), Left(bFailures)) => Left(aFailures ++ bFailures)
+      case (_, l: Left[_, _]) => l.asInstanceOf[Left[ConfigReaderFailures, Nothing]]
+      case (l: Left[_, _], _) => l.asInstanceOf[Left[ConfigReaderFailures, Nothing]]
+    }
+
+  def fail[A](failure: ConfigReaderFailure): ConfigReaderResult[A] = Left(ConfigReaderFailures(failure))
+
+  def failWithThrowable[A](throwable: Throwable): ConfigReaderResult[A] = fail[A](ThrowableFailure(throwable))
+
+  private def eitherToResult[T](either: Either[ConfigReaderFailure, T]): ConfigReaderResult[T] =
+    either match {
+      case r: Right[_, _] => r.asInstanceOf[ConfigReaderResult[T]]
+      case Left(failure) => Left(ConfigReaderFailures(failure))
+    }
+
+  private def tryToEither[T](t: Try[T]): Either[ConfigReaderFailure, T] = t match {
+    case Success(t) => Right(t)
+    case Failure(e) => Left(ThrowableFailure(e))
+  }
+
+  private def stringToTryConvert[T](fromF: String => Try[T]): ConfigValue => ConfigReaderResult[T] =
+    stringToEitherConvert[T](fromF andThen tryToEither[T])
+
+  private def stringToEitherConvert[T](fromF: String => Either[ConfigReaderFailure, T]): ConfigValue => ConfigReaderResult[T] =
     config => {
-      // Because we can't trust `fromF` or Typesafe Config not to throw, we wrap the
-      // evaluation in one more `Try` to prevent an unintentional exception from escaping.
-      // `Try.flatMap(f)` captures any non-fatal exceptions thrown by `f`.
-      Try(config.valueType match {
-        case ConfigValueType.STRING => config.unwrapped.toString
-        case _ => config.render(ConfigRenderOptions.concise)
-      }).flatMap(fromF)
+      // Because we can't trust Typesafe Config not to throw, we wrap the
+      // evaluation into a `try-catch` to prevent an unintentional exception from escaping.
+      try {
+        val string = config.valueType match {
+          case ConfigValueType.STRING => config.unwrapped.toString
+          case _ => config.render(ConfigRenderOptions.concise)
+        }
+        eitherToResult(fromF(string))
+      } catch {
+        case NonFatal(t) => failWithThrowable(t)
```

Comment:
This might be out of the scope of this PR, but I believe we can provide more context here, namely about the type we were trying to convert to and possibly about the `ConfigValue` we tried to convert from.